icu4c/source/i18n/measunit_extra.cpp

```
@@ -530,8 +548,116 @@ class Token {
+                    return result;
+                }
+                continue;
+            }
+
+            // Handle 'e' or 'E'
+            if (c == 'e' || c == 'E') {
+                exponentIndex = i + 1;
+                break;
+            }
+
+            // Invalid character
+            status = kUnitIdentifierSyntaxError;
+            return result;
+        }
+
+        // Handle exponent
+        for (int i = exponentIndex; i < str.length(); ++i) {
+            char c = str.data()[i];
+
+            // handle sign
+            if (i == exponentIndex && c == '+') {
+                continue; // Skip leading plus sign
+            }
+
+            if (i == exponentIndex && c == '-') {
+                // Negative sign is not allowed for the exponent.
+                status = kUnitIdentifierSyntaxError;
+                return result;
+            }
+
+            // Handle digits
+            if (c >= '0' && c <= '9') {
+                uint16_t digit = c - '0';
+                exponent = exponent * 10 + digit;
+
+                // Check if the exponent is within the valid range (0 to int32_t::MAX)
+                if (exponent > 18) {
+                    status = kUnitIdentifierSyntaxError;
+                    return result;
+                }
+
+                continue;
+            }
+
+            // Invalid character
+            status = kUnitIdentifierSyntaxError;
+            return result;
+        }
+
+        // Apply the exponent
+        for (int32_t i = 0; i < exponent; ++i) {
+            result *= 10;
+
+            if (result > max_int64) {
+                status = kUnitIdentifierSyntaxError;
+                return result;
+            }
+        }
+
+        // Check if the result is within the valid range (1 to max_int64)
+        if ( result < 1 || result > max_int64) {
+            status = kUnitIdentifierSyntaxError;
+            return result; // add it for code consistency.
+        }
+
+        return result;
+    }
```

Two comments about this whole function:

1) Since this is guaranteed to be a decimal number using ASCII characters in a `char` string, couldn't you just use `atoi()` instead of writing the whole thing yourself?  Afterward, you could just check the result from `atoi()` to decide whether it's valid (e.g., disallowing negative numbers).

2) The logic to parse the mantissa and the logic to parse the exponent are basically the same (except maybe for which values are valid).  It seems like you could use the same code to handle each piece.  In fact, if `atoi()` doesn't support scientific notation, you could at least use it to separately parse the mantissa and exponent and then just have your own logic for handling the `e` and putting the pieces together (or use `atof()`, although that'll give you back a `float`, which probably isn't helpful).

3) Come to think of it, I don't think the logic above works right with EBCDIC, although I don't know if we still care about that...

---

I have used StringToDoubleConverter for now to mimic the same behavior in Java. However, we need to implement our own logic in a separate class to return an error when there is a limit exceed instead of returning any random number.

I am trying to implement the long converter here: https://github.com/unicode-org/icu/pull/3339